### ISSUE TITLE
Compiler runs out of memory when this types are added to `Array<T>` methods

### ISSUE THREAD

**Issue by sandersn** [Microsoft/TypeScript#9707]
**TypeScript Version:**  master

**Code**

1. In es5.d.ts, add `this: this` parameters to all Array methods
2. Run inferringAnyFunctionType2 (reproduced below)

```ts
function f<T extends [(p1: number) => number]>(p: T): T {
    return p;
}

var v = f([x => x]);
```

**Expected behavior:**
The test passes

**Actual behavior:** 
TypeScript runs out of memory.

It looks as if something about instantiating `this` types in a constraint evades the relation cache and keeps instantiating types. This might have something to do with the special handling of arrays and tuples, because I can't get the failure to repro when switching to a non built-in type.

**Comment by sandersn** [comment]
Here is a self-contained repro thanks to @vladima, who showed me how to modify Array in a test file:

```ts
/// <reference no-default-lib="true"/>

interface Boolean {}
interface IArguments {}
interface Function {}
interface Number {}
interface RegExp {}
interface Object {}
declare class String { charCodeAt(a: number): number }

interface Array<T> {
    map<U>(arg: this): void;
    reduceRight<U>(arg: this): void;
    reduce<U>(arg: this): void;
    reduce2<U>(arg: this): void;
}

declare function f<T extends [(x: number) => number]>(a: T): void;
f([String.prototype.charCodeAt])
```

The bug still happens when you comment out all but one of the remaining Array methods, but it doesn't run out of memory until there are four of them.

**Comment by sandersn** [comment]
Fix is up at #10234
